### PR TITLE
Get config values from BTCD #31

### PR DESCRIPTION
**Pull request by ImmanuelSegol**
Please review, and comment on what I should add/remove/improve.
Note that the btcd.conf is not in toml format so I had to parse it my self.

### CODE REVIEW

**Comment by michael1011** [comment]
@ImmanuelSegol Yeah I was wrong about that. The config file of BTCD is an [INI](https://en.wikipedia.org/wiki/INI_file) one. Maybe you can find a library for parsing it so that we don't have to do it manually.

File: lib/Config.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        this.btcdDir = `${localDir}/Btcd/`;
+        break;
+      }
+      case 'darwin': {
+        const localDir = process.env.HOME;
+        this.walliDir = `${localDir}/.walli/`;
+        this.btcdDir = `${localDir}/Btcd/`;
+        break;
+      }
+      default: /* WIN32 */ {
```
Comment (michael1011) [inline lib/Config.ts]: This is a minor detail but FreeBSD and OpenBSD would be treated like a Windows operating system. Use `win32` to detect Windows like systems and use `default` for Linux and the BSD ones.

File: lib/Walli.ts [hunk 2]
```diff
@@ -12,10 +13,13 @@ class Walli {
...
   private logger: Logger;
   private btcdClient: BtcdClient;
+  private config: Config;
 
-  constructor(config: Config) {
-    this.logger = new Logger(config.logfile, config.loglevel);
-    this.btcdClient = new BtcdClient(config.rpc);
+  constructor(config: ConfigType) {
+    this.config = new Config().load(config);
+    console.log(this.config);
```
Comment (michael1011) [inline lib/Walli.ts]: Please remove this `console.log` statement in the next commit. And what code editor are you using? If it is Visual Studio Code I could commit my configurations for debugging.

Comment (ImmanuelSegol) [reply lib/Walli.ts]: Oops didn't notice that and use I'm using VS Code.

File: lib/Walli.ts [hunk 3]
```diff
@@ -29,6 +32,7 @@ class Walli {
       this.logger.error(`could not connect to BTCD: ${JSON.stringify(error)}`);
     }
   }
+
```
Comment (michael1011) [inline lib/Walli.ts]: Please remove this blank line

File: lib/Config.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        const localDir = process.env.LOCALAPPDATA;
+        this.btcdDir = `${localDir}/Btcd/`;
+        break;
+      }
+      case 'darwin': {
+        const localDir = process.env.HOME;
+        this.btcdDir = `${localDir}/Btcd/`;
+        break;
+      }
+      default: /* WIN32 */ {
```
Comment (michael1011) [inline lib/Config.ts]: Remove this comment as it is not true anymore

File: lib/Config.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  public logfile: string;
+  public loglevel: string;
+  public rpc: RpcConfig;
+  private btcdDir: string;
+
+  constructor() {
+    const platform = os.platform();
+
+    switch (platform){
+      // TODO: Check if the btcd.conf file directory is proper for linux and mac
```
Comment (michael1011) [inline lib/Config.ts]: For Linux it is `process.env.HOME/.btcd` and Mac is `process.env.HOME/Library/Application Support/Btcd` (but I am not sure).

File: lib/Config.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    if (args) {
+      deepMerge(this, args);
+    }
+
+    return this;
+  }
+
+  private loadBtcdRpcConfig (path: string): object {
```
Comment (michael1011) [inline lib/Config.ts]: We could parse all values (even port and host) of the BTCD config and use the default ones if nothing is set in the config. Then we could return `RpcConfig`

File: lib/Config.ts [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    this.btcdconfig = 'btcd.conf';
+    this.config = 'walli.conf';
+    this.loglevel = 'verbose';
+  }
+
+  public load(args?: { [argName: string]: any }) {
+    const walliPath = path.join(__dirname, '../', this.config);
+    const btcdPath = path.join(this.btcdDir, this.btcdconfig);
+
+    if (fs.existsSync(this.btcdDir)) {
```
Comment (michael1011) [inline lib/Config.ts]: We don't really care about the directory but just about the BTCD config file itself. (this should be: `if (fs.existsSync(btcdPath)) {`)

File: .gitignore [hunk 8]
```diff
@@ -64,4 +64,5 @@ dist/
 typedoc/
 
 # Config file
-walli.json
+walli.conf
+
```
Comment (michael1011) [inline .gitignore]: Please remove this blank line.

File: lib/Config.ts [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+    this.logfile = 'walli.log';
+    this.btcdconfig = 'btcd.conf';
+    this.config = 'walli.conf';
+    this.loglevel = 'verbose';
+  }
+
+  public load(args?: { [argName: string]: any }) {
+    const walliPath = path.join(__dirname, '../', this.config);
+    const btcdPath = path.join(this.btcdDir, this.btcdconfig);
```
Comment (michael1011) [inline lib/Config.ts]: You should check if `args` contains values for the config files first.

File: lib/Config.ts [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+      const walliTOML = fs.readFileSync(walliPath, 'utf-8');
+
+      try {
+        const confFile = toml.parse(walliTOML);
+        deepMerge(this, confFile);
+      } catch (e) {
+        throw new Error(`Error on line ${e.line}, cloumn ${e.cloumn}: ${e.message}`);
+      }
+    } else {
+      fs.writeFileSync(this.config, '');
```
Comment (michael1011) [inline lib/Config.ts]: Why should we create an empty config file?

File: lib/Config.ts [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    if (fs.existsSync(walliPath)) {
+      const walliTOML = fs.readFileSync(walliPath, 'utf-8');
+
+      try {
+        const confFile = toml.parse(walliTOML);
+        deepMerge(this, confFile);
+      } catch (e) {
+        throw new Error(`Error on line ${e.line}, cloumn ${e.cloumn}: ${e.message}`);
```
Comment (michael1011) [inline lib/Config.ts]: What about `Could not parse BTCD config. Using default values` as error? BTCD will complain about the error anyways therefore we don't have to care about it.

**Pull request merged**